src/libAtomVM/posix_nifs.c

```diff
@@ -488,6 +490,130 @@ static term nif_atomvm_posix_select_stop(Context *ctx, int argc, term argv[])
 
     return OK_ATOM;
 }
+
+#if HAVE_EXECVE
+static char **parse_string_list(term list)
+{
+    if (!term_is_list(list)) {
+        return NULL;
+    }
+    int ok;
+    size_t result_len = term_list_length(list, &ok);
+    if (UNLIKELY(!ok)) {
+        return NULL;
+    }
+    // All items are initialized to NULL.
+    char **result_list = calloc(result_len + 1, sizeof(char *));
+    if (IS_NULL_PTR(result_list)) {
+        return NULL;
+    }
+    term list_item = list;
+    int i = 0;
+    while (term_is_nonempty_list(list_item)) {
+        term item = term_get_list_head(list_item);
+        char *str = interop_term_to_string(item, &ok);
+        if (UNLIKELY(!ok)) {
+            for (int j = 0; j < i; j++) {
+                free(result_list[j]);
+            }
+            free(result_list);
+            return NULL;
+        }
+        result_list[i++] = str;
+        list_item = term_get_list_tail(list_item);
+        if (str == NULL) {
+            // NULL pointer marks the end of the list, so we need to break
+            // here.
+            break;
+        }
+    }
+    return result_list;
+}
+
+static void free_string_list(char **list)
+{
+    char **ptr = list;
+    while (*ptr) {
+        char *str = *ptr;
+        free(str);
+        ptr++;
+    }
+    free(list);
+}
+
+static term nif_atomvm_subprocess(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+
+    int ok;
+    char *path = interop_term_to_string(argv[0], &ok);
+    if (UNLIKELY(!ok)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    char **args = parse_string_list(argv[1]);
+    if (IS_NULL_PTR(args)) {
+        free(path);
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    char **envp = parse_string_list(argv[2]);
+    if (IS_NULL_PTR(envp)) {
+        free(path);
+        free_string_list(args);
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    int pstdout[2];
+    int r = pipe(pstdout);
+    if (r < 0) {
+        free(path);
+        free_string_list(args);
+        free_string_list(envp);
+        return errno_to_error_tuple_maybe_gc(ctx);
+    }
+    r = fork();
+    if (r < 0) {
+        free(path);
+        free_string_list(args);
+        free_string_list(envp);
+        close(pstdout[0]);
+        close(pstdout[1]);
+        return errno_to_error_tuple_maybe_gc(ctx);
+    }
+    if (r == 0) {
+        // child.
+        close(0); // close stdin of the child
+        close(pstdout[0]); // close read end of the pipe
+        dup2(pstdout[1], 1); // make stdout the write-end of the pipe
+        close(pstdout[1]); // close excess file descriptor
```

Comment:
Sadly it is not so simple, this will leak a lot of other file descriptors in the newly created process.
We need to close all open files, such as the beam modules and the sockets we have around.
The safest way around is using `O_CLOEXEC` and similar flags in order to have all file descriptors around closed on execve.
Sadly, `O_CLOEXEC` is part of newer POSIX standard versions, so my suggestion here is making this specific NIF available only when `O_CLOEXEC` is available.
But maybe there are other possible approaches.

Reply:
Sounds like a good approach.